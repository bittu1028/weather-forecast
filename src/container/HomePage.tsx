import React, { useCallback, useEffect, useState } from "react";
import Header from "../components/header/Header";
import Weather from "../components/weather/Weather";
import Forecast from "../components/forecast/Forecast";
import Loader from "../shared/loader/Loader";
import { ErrorBoundary } from "react-error-boundary";
import { Error, ErrorHandler } from "../shared/error/Error";
import { useDispatch, useSelector } from "react-redux";
import { AppDispatch, RootState } from "../reducers/store";
import { fetchWeather } from "../reducers/fetchWeather";
import { NotFoundError } from "../shared/error/NotFound";

function HomePage() {
  const dispatch = useDispatch<AppDispatch>();
  const [searchQuery, setSearchQuery] = useState<string>("");
  const [isCelcius, setIsCelcius] = useState<boolean>(false);
  const { 
    currentWeather, 
    forecast, 
    isLoading, 
    cityInfo, 
    isError,
    errorMessage 
  } =
  useSelector((state: RootState) => state.weather);

  const onSearchQueryChange = (e: any) => {
    const input = e.target.value.toLowerCase().trimStart();
    setSearchQuery(input);
  };

  const getWeatherData = useCallback(async (location: string) => {
    dispatch(fetchWeather(location));
  }, []);

  useEffect(() => {
    getWeatherData("Sydney");
  }, []);

  const onSearchWeather = () => {
    if (searchQuery.length !== 0 && !isLoading) {
      getWeatherData(searchQuery);
    }
  };

  const onKeyEnter = (e: any) => {
    if (e.keyCode === 13) {
      onSearchWeather();
    }
  };

  const onToggleFahrenheit = () => {
    setIsCelcius(!isCelcius);
  };

  return (
    <>
      <ErrorBoundary FallbackComponent={Error} onError={ErrorHandler}>
        {isLoading && <Loader />}
        <Header
          isSearching={isLoading}
          onChange={onSearchQueryChange}
          onKeyDown={onKeyEnter}
          onSearchWeather={onSearchWeather}
          searchQuery={searchQuery}
        />
        <div className="app-content">
          {
            isError ? <NotFoundError errorMessage={errorMessage}/> : null
          }
          {!isLoading && currentWeather && cityInfo && !isError && (
            <React.Fragment>
              <Weather
                onToggle={onToggleFahrenheit}
                weather={currentWeather}
                cityInfo={cityInfo}
              />
              {forecast.length !== 0 && <Forecast forecast={forecast} />}
            </React.Fragment>
          )}
        </div>
      </ErrorBoundary>
    </>
  );
}

export default HomePage;
