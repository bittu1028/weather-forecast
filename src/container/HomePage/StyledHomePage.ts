import styled from 'styled-components';
import { device } from '../../utils/device';

export const StyledContent = styled.div`
  width: 70%;
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
  margin: auto;
  padding: $xl-size;
  text-align: left;

  @media ${device.tablet} {
    width: 100%;
  }

  @media ${device.mobile} {
    padding: ${(props) => props.theme.ls};
    padding-bottom: ${(props) => props.theme.xl};
  }
`;
