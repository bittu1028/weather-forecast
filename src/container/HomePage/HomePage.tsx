import React, { useCallback, useEffect, useState } from 'react';
import Header from '../../components/header/Header';
import Weather from '../../components/weather/Weather';
import Forecast from '../../components/forecast/Forecast';
import Loader from '../../shared/loader/Loader';
import { ErrorBoundary } from 'react-error-boundary';
import { Error, ErrorHandler } from '../../shared/error/Error';
import { useDispatch, useSelector } from 'react-redux';
import { AppDispatch, RootState } from '../../redux/store';
import { fetchWeather } from '../../redux/fetchWeather';
import { NotFoundError } from '../../shared/error/NotFound';
import { StyledContent } from './StyledHomePage';
import Units from '../../components/units/Units';

const currentCity = 'Sydney'

const HomePage = () => {
  const dispatch = useDispatch<AppDispatch>();
  // state for search query
  const [searchQuery, setSearchQuery] = useState<string>('');
  // to toggle format of temperature
  const [unit, setUnit] = useState<string>('metric');
  // reducers data
  const { currentWeather, forecast, isLoading, cityInfo, isError, errorMessage } = useSelector(
    (state: RootState) => state.weather
  );

  const onSearchQueryChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const input = e.target.value.toLowerCase().trimStart();
    setSearchQuery(input);
  };

  const getWeatherData = useCallback(async (location: string, currentUnit:string = unit) => {
    // dispatching action to fetch data from open weather
    dispatch(fetchWeather({location, currentUnit}));
  }, []);

  useEffect(() => {
    // defaulting value to sydney
    // future enhanncements to use current location
    getWeatherData(currentCity);
  }, []);

  const onSearchWeather = () => {
    if (searchQuery.length !== 0 && !isLoading) {
      getWeatherData(searchQuery);
    }
  };

  // call api if user press enter key on search box
  const onKeyEnter = (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.keyCode === 13) {
      onSearchWeather();
    }
  };
  
  const changeUnitSettings = (currentUnit:string) => {
    setUnit(currentUnit);
    getWeatherData(searchQuery || currentCity, currentUnit);
  };

  return (
    <>
      <ErrorBoundary FallbackComponent={Error} onError={ErrorHandler}>
        {isLoading && <Loader />}
        <Header
          isSearching={isLoading}
          onChange={onSearchQueryChange}
          onKeyDown={onKeyEnter}
          onSearchWeather={onSearchWeather}
          searchQuery={searchQuery}
        />
        <StyledContent>
          <Units 
              changeUnitSettings={changeUnitSettings}
              unit={unit}
          />
          {isError ? <NotFoundError errorMessage={errorMessage} /> : null}
          {!isLoading && currentWeather && cityInfo && !isError && (
            <React.Fragment>
              <Weather weather={currentWeather} cityInfo={cityInfo} />
              {forecast.length !== 0 && <Forecast forecast={forecast} />}
            </React.Fragment>
          )}
        </StyledContent>
      </ErrorBoundary>
    </>
  );
}

export default HomePage;
