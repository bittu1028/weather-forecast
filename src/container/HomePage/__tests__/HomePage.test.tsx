// __tests__/HomePage.test.tsx
import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import HomePage from '../HomePage';
import store from '../../../reducers/store';

const renderHomePage = () =>
  render(
    <Provider store={store}>
      <HomePage />
    </Provider>
  );

describe('HomePage', () => {
  test('renders current homepage component', async () => {
    renderHomePage();
    const titleInput = screen.getByPlaceholderText('Search for city');
    expect(titleInput).toBeInTheDocument();
  });
});

describe('HomePage', () => {
  test('renders current homepage component', async () => {
    const { asFragment } = renderHomePage();
    expect(asFragment).toMatchSnapshot();
  });
});
