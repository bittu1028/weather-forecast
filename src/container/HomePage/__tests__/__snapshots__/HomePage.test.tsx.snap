// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage renders current homepage component 1`] = `[Function]`;
